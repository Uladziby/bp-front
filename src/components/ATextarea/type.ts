export interface IATextarea {
  placeholder: string;
  label: string;
  hasError?: boolean;
  isDisabled?: boolean;
  name: string;
  defaultValue?: string;
  maxLength?: number;
  rows?: number;
}
