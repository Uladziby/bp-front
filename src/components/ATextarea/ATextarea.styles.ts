/** @format */

import styled from "styled-components";
import { COLOR_PALETTE, INPUT_COLORS } from "utils/colors";

export const StyledWrapper = styled.div`
	position: relative;
	display: flex;
	flex-direction: column;
	height: 100%;
	margin: 1rem 2rem;
`;

export const StyledLabel = styled.label`
	cursor: pointer;
	text-align: left;
	color: ${INPUT_COLORS.default};
	font-size: 14px;
`;

export const StyledTextarea = styled.textarea<{ $hasError?: boolean }>`
	margin: 8px 0;
	height: 100%;
	line-height: 22px;
	padding: 1rem;
	padding-right: ${({ $hasError }) => $hasError && 34}px;
	resize: none;
	border: 1px solid ${INPUT_COLORS.default};
	border-radius: 4px;

	&:focus {
		box-shadow: 0px 0px 0px 1px ${INPUT_COLORS.default} inset;
		outline: none;
	}

	&:focus::placeholder {
		color: transparent;
	}
`;

export const StyledErrorIcon = styled.div`
	position: absolute;
	top: 50%;
	right: 12px;
	transform: translateY(-50%);
`;

export const StyledCounter = styled.span`
	align-self: flex-end;
	font-size: 12px;
	color: ${COLOR_PALETTE.gray900};
`;

export const StyledErrorMessage = styled.div`
	width: 100%;
	text-align: left;
	color: ${INPUT_COLORS.error};
	font-size: 12px;
`;
