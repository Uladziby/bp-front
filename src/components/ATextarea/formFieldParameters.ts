export const textareaLineHeight = 22;
export const textareaVerticalPaddings = 22;
export const borderWidth = 2;
