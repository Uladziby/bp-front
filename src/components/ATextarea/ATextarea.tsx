/** @format */

import { useEffect, useRef, useState } from "react";
import { useController } from "react-hook-form";
import { ReactComponent as DangerIcon } from "assets/icons/inputAlert.svg";
import {
	StyledCounter,
	StyledErrorIcon,
	StyledErrorMessage,
	StyledLabel,
	StyledTextarea,
	StyledWrapper,
} from "./ATextarea.styles";
import { IATextarea } from "./type";
import { textareaLineHeight, textareaVerticalPaddings, borderWidth } from "./formFieldParameters";

export const ATextarea = ({
	name,
	placeholder,
	label,
	isDisabled,
	maxLength,
	rows = 1,
	defaultValue,
}: IATextarea) => {
	const [isShowCounter, setIsShowCounter] = useState(false);

	const textAreaRef = useRef<HTMLTextAreaElement>(null);

	const handleFocus = () => {
		if (maxLength) {
			setIsShowCounter((prevState) => !prevState);
		}
	};

	const {
		field,
		fieldState: { error },
	} = useController({ name, shouldUnregister: true, defaultValue: defaultValue || "" });

	useEffect(() => {
		if (!textAreaRef.current) return;

		const initTextareaHeight = rows * textareaLineHeight + textareaVerticalPaddings;

		textAreaRef.current.style.height = initTextareaHeight + "px";

		textAreaRef.current.style.height = textAreaRef.current.scrollHeight + borderWidth + "px";
	}, [field.value, rows]);

	return (
		<StyledWrapper onFocus={handleFocus} onBlur={handleFocus}>
			<StyledLabel htmlFor={name}>{label}</StyledLabel>
			<StyledTextarea
				{...field}
				id={name}
				placeholder={placeholder}
				disabled={isDisabled}
				rows={rows}
				$hasError={Boolean(error)}
				ref={textAreaRef}
			/>
			{error && (
				<StyledErrorIcon>
					<DangerIcon />
				</StyledErrorIcon>
			)}
			{field.value && !error && isShowCounter && (
				<StyledCounter data-testid="counter">
					{field.value.trim().length}/{maxLength}
				</StyledCounter>
			)}
			{error && <StyledErrorMessage>{String(error.message)}</StyledErrorMessage>}
		</StyledWrapper>
	);
};
