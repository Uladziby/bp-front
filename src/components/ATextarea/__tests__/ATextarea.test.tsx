import { RenderResult, render, renderHook, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { FormProvider, useForm } from 'react-hook-form';
import { INPUT_COLORS } from 'constants/colors/colors';
import { ATextarea } from '../ATextarea';

describe('ATextarea', () => {
  const { result } = renderHook(() => useForm());
  const form = result.current;

  const props = {
    name: 'name',
    label: 'label',
    placeholder: 'placeholder',
    isDisabled: false,
    defaultValue: 'value',
    rows: 2,
    maxLength: 20,
  };
  let renderResult: RenderResult;

  beforeEach(() => {
    renderResult = render(
      <FormProvider {...form}>
        <ATextarea {...props} />
      </FormProvider>,
    );
  });

  it('should render correctly', () => {
    const textarea = screen.getByRole('textbox');

    expect(textarea).toBeInTheDocument();
  });

  it('snapshot', () => {
    const textarea = screen.getByRole('textbox');
    expect(textarea).toMatchSnapshot();
  });

  it('handles property name correctly', () => {
    const labelEl = screen.getByText(props.label);
    const textarea = screen.getByLabelText(props.label);

    expect(labelEl).toHaveAttribute('for', props.name);
    expect(textarea).toHaveAttribute('id', props.name);
  });

  it('handles property label correctly', () => {
    const label = screen.getByText(props.label);

    expect(label.tagName).toBe('LABEL');
    expect(label.textContent).toBe(props.label);
  });

  it('handles property placeholder correctly', () => {
    const textarea = screen.getByPlaceholderText(props.placeholder);

    expect(textarea).toBeInTheDocument();
  });

  it('handles property isDisabled=true correctly', () => {
    renderResult.rerender(
      <FormProvider {...form}>
        <ATextarea {...props} isDisabled />
      </FormProvider>,
    );

    const textarea = screen.getByRole('textbox');

    expect(textarea).toHaveStyle({
      border: `1px solid ${INPUT_COLORS.disabled}`,
    });
  });

  it('handles property isDisabled=false correctly', () => {
    const textarea = screen.getByRole('textbox');

    expect(textarea).toHaveStyle({ border: `1px solid ${INPUT_COLORS.default}` });
  });

  it('handles property defaultValue correctly', () => {
    const textarea = screen.getByRole('textbox');

    expect(textarea.textContent).toBe(props.defaultValue);
  });

  it('handles property maxLength correctly', () => {
    const value = 'some value';

    const textarea = screen.getByRole('textbox');

    userEvent.click(textarea);
    userEvent.type(textarea, value);

    const counter = screen.getByTestId('counter');

    expect(counter.textContent).toBe(
      `${value.length + props.defaultValue.length}/${props.maxLength}`,
    );
  });
});
