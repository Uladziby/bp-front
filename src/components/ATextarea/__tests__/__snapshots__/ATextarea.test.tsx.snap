// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ATextarea snapshot 1`] = `
<textarea
  class="sc-dkrFOg bDjIvE"
  id="name"
  name="name"
  placeholder="placeholder"
  rows="2"
  style="height: 2px;"
>
  value
</textarea>
`;
